### PR TITLE
EIP-98: Optional transaction state root

### PR DESCRIPTION
**Pull request by arkpar**
This PR removes intermediate trie commits when EIP-98 is enabled. Actual transaction parallelization is to follow.

### CODE REVIEW

File: ethcore/src/block.rs [hunk 1]
```diff
@@ -375,6 +375,9 @@ impl<'x> OpenBlock<'x> {
 		let unclosed_state = s.block.state.clone();
 
 		s.engine.on_close_block(&mut s.block);
+		if let Err(e) = s.block.state.commit() {
+			warn!("Encountered error on state commit: {}", e);
```
Comment (tomusdrw) [inline ethcore/src/block.rs]: Shouldn't that error be propagated upstream?

File: ethcore/src/spec/spec.rs [hunk 2]
```diff
@@ -65,6 +67,7 @@ impl From<ethjson::spec::Params> for CommonParams {
 			subprotocol_name: p.subprotocol_name.unwrap_or_else(|| "eth".to_owned()),
 			min_gas_limit: p.min_gas_limit.into(),
 			fork_block: if let (Some(n), Some(h)) = (p.fork_block, p.fork_hash) { Some((n.into(), h.into())) } else { None },
+			eip98_transition: p.eip98_transition.map_or(0, Into::into),
```
Comment (tomusdrw) [inline ethcore/src/spec/spec.rs]: Wouldn't that enable `eip98` automatically for spec files where that prop is not defined (private chains?)?

File: ethcore/src/state/mod.rs [hunk 3]
```diff
@@ -519,8 +519,14 @@ impl State {
 
 		// TODO uncomment once to_pod() works correctly.
 //		trace!("Applied transaction. Diff:\n{}\n", state_diff::diff_pod(&old, &self.to_pod()));
-		self.commit()?;
-		let receipt = Receipt::new(self.root().clone(), e.cumulative_gas_used, e.logs);
+		let state_root = match env_info.number < engine.params().eip98_transition {
```
Comment (tomusdrw) [inline ethcore/src/state/mod.rs]: Using `if-else` here would a bit more idiomatic and suprisingly shorter then `match`.

File: ethcore/src/block.rs (hunk 1 above)
Comment (arkpar) [inline ethcore/src/block.rs]: This is how commit error are handles in the rest of the codebase.

File: ethcore/src/spec/spec.rs (hunk 2 above)
Comment (arkpar) [inline ethcore/src/spec/spec.rs]: Yes, that is the intention

**Pull request merged**